component/kafka/component.go

```diff
@@ -0,0 +1,427 @@
+	log.Debug("kafka component: consumer ready")
+	<-ctx.Done()
+	log.Infof("kafka component terminating: context cancelled")
+	wg.Wait()
+	err = client.Close()
+	if err != nil {
+		return err
+	}
+
+	return handler.err
+}
+
+// Consumer represents a sarama consumer group consumer
+type consumerHandler struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	name  string
+	group string
+	ready chan bool
+
+	// buffer
+	batchSize int
+	ticker    *time.Ticker
+	msgBuf    []*sarama.ConsumerMessage
+
+	// retries
+	retries   int
+	retryWait time.Duration
+
+	// lock to protect buffer operation
+	mu sync.RWMutex
+
+	// callback
+	proc BatchProcessorFunc
+
+	// failures strategy
+	failStrategy FailStrategy
+
+	// processing error
+	err error
+}
+
+func newConsumerHandler(ctx context.Context, cancel context.CancelFunc, name, group string, processorFunc BatchProcessorFunc,
+	fs FailStrategy, batchSize uint, batchTimeout time.Duration, retries uint, retryWait time.Duration) *consumerHandler {
+
+	return &consumerHandler{
+		ctx:          ctx,
+		cancel:       cancel,
+		name:         name,
+		group:        group,
+		batchSize:    int(batchSize),
+		ready:        make(chan bool),
+		ticker:       time.NewTicker(batchTimeout),
+		msgBuf:       make([]*sarama.ConsumerMessage, 0, batchSize),
+		retries:      int(retries),
+		retryWait:    retryWait,
+		mu:           sync.RWMutex{},
+		proc:         processorFunc,
+		failStrategy: fs,
+	}
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (c *consumerHandler) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	close(c.ready)
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (c *consumerHandler) Cleanup(sarama.ConsumerGroupSession) error {
+	// if the strategy is to exit on failure then we cancel the context to exit the whole component
+	// and thus prevent the Consume loop from picking up the message again
+	if c.failStrategy == ExitStrategy {
+		c.cancel()
+	}
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (c *consumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for {
+		select {
+		case msg, ok := <-claim.Messages():
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, topic = %s", string(msg.Value), msg.Timestamp, msg.Topic)
+				topicPartitionOffsetDiffGaugeSet(c.group, msg.Topic, msg.Partition, claim.HighWaterMarkOffset(), msg.Offset)
+				messageStatusCountInc(messageReceived, c.group, msg.Topic)
+				err := c.insertMessage(session, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			c.mu.Lock()
+			err := c.flush(session)
+			if err != nil {
+				return err
+			}
+			c.mu.Unlock()
+		case <-c.ctx.Done():
+			if c.ctx.Err() != context.Canceled {
```

Comment:
It can be also a cancellation from a parent context, I am afraid we should always log it so that such specific situation (app parent context cancelled) can be tracked down